plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -853,7 +853,25 @@ public void getUCJson(VersionNumber jenkinsVersion) {
 
         String cacheSuffix = getCacheSuffix(jenkinsVersion);
         try {
-            URIBuilder uriBuilder = new URIBuilder(cfg.getJenkinsUc().toURI());
+
+            URL ucUrl = cfg.getJenkinsUc();
+
+            if (cfg.isUseArchiveAll()) {
+                URL archiveUcUrl = resolveArchiveUpdateCenterUrl(cacheSuffix, cfg.getJenkinsArchiveRepoMirror());
+                if (archiveUcUrl != null) {
+                    ucUrl = archiveUcUrl;
+                } else {
+                    logVerbose("No valid archive repository found for version " + jenkinsVersion + ". Falling back to the default repository.");
+                }
+            }
+
+            if (ucUrl == null) {
+                throw new IllegalStateException("No valid update center URL found for version " + jenkinsVersion);
+            }
+
+            URI ucUri = ucUrl.toURI();
+
+            URIBuilder uriBuilder = new URIBuilder(ucUri);
```

Comment:
Could most of this be extracted to a method to hide away some of the logic which isn't critical most of the time to the reader?